openstack/messaging/v2/queues/requests.go

```diff
@@ -0,0 +1,65 @@
+package queues
+
+import (
+	"github.com/gophercloud/gophercloud"
+)
+
+// CreateOptsBuilder allows extensions to add additional parameters to the
+// Create request.
+type CreateOptsBuilder interface {
+	ToQueueCreateMap() (map[string]interface{}, error)
+}
+
+// CreateOpts specifies the queue creation parameters.
+type CreateOpts struct {
+	//The target incoming messages will be moved to when a message can’t
+	// processed successfully after meet the max claim count is met.
+	DeadLetterQueue string `json:"_dead_letter_queue,omitempty"`
+
+	// The new TTL setting for messages when moved to dead letter queue.
+	DeadLetterQueueMessageTTL int `json:"_dead_letter_queue_messages_ttl,omitempty"`
+
+	// The delay of messages defined for a queue. When the messages send to
+	// the queue, it will be delayed for some times and means it can not be
+	// claimed until the delay expired.
+	DefaultMessageDelay int `json:"_default_message_delay,omitempty"`
+
+	// The default TTL of messages defined for a queue, which will effect for
+	// any messages posted to the queue.
+	DefaultMessageTTL int `json:"_default_message_ttl" required:"true"`
+
+	// The flavor name which can tell Zaqar which storage pool will be used
+	// to create the queue.
+	Flavor string `json:"_flavor,omitempty"`
+
+	// The max number the message can be claimed.
+	MaxClaimCount int `json:"_max_claim_count,omitempty"`
+
+	// The max post size of messages defined for a queue, which will effect
+	// for any messages posted to the queue.
+	MaxMessagesPostSize int `json:"_max_messages_post_size,omitempty"`
+
+	// Description of the queue.
+	Description string `json:"description,omitempty"`
+}
+
+// ToQueueCreateMap constructs a request body from CreateOpts.
+func (opts CreateOpts) ToQueueCreateMap() (map[string]interface{}, error) {
+	return gophercloud.BuildRequestBody(opts, "")
+}
+
+// Create requests the creation of a new queue.
+func Create(client *gophercloud.ServiceClient, queueName string, clientID string, opts CreateOptsBuilder) (r CreateResult) {
```

Comment:
I'm not a fan of this function signature -- there's a lot going on.

An alternative might be to add `QueueName` and `ClientID` to the `CreateOpts` struct. Then below, you could do:

```go
b, err := opts.ToQueueCreateMap()
...

clientID := b["client_id"]
delete(b, "client_id")
queueName := b["queue_name"]
delete(b, "queue_name")

...
```

Alternatively, if `clientID` is going to be a staple in all headers, then maybe only remove `queueName` from the signature.

Having `QueueName` in the `CreateOpts` makes sense since that follows patterns seen throughout the rest of Gophercloud. For example, when creating a user, the username is part of CreateOpts. I realize that the queue name is part of the request URL here, but as shown above, we can work around that.

The reason I'm focusing on this is because if you look at all `Create` function signatures in Gophercloud, the vast majority only have 2 parameters: client and createopts. A handful have a third parameter for a parent resource ID. Unless we _really_ need to, I'd like to keep this close to form.

Thoughts?

Reply:
I agree with this 100%. 

ClientID is a required header for (almost) all requests so it makes sense to leave it in the function signature IMO but I will remove queueName. 

Unless you feel really strongly about moving clientID into createOpts I would prefer to leave it as (almost) all of the other functions will have clientID in the signature.

Reply:
> ClientID is a required header for (almost) all requests so it makes sense to leave it in the function signature IMO but I will remove queueName.

This sounds good to me. If client ID will be used for most/all other functions, having it consistently in the function signature is a good idea.